File: app/src/main/java/com/example/digitalmonk/service/vpn/blocklist/BlocklistManager.java
package com.example.digitalmonk.service.vpn.blocklist;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Why we made this file:
 * A DNS-level VPN needs to know exactly which websites to block in real-time.
 * Checking a database for every single network packet is too slow. This manager
 * loads all blocked domains (from local seeds, cached files, and remote servers)
 * directly into device RAM (Memory) for lightning-fast O(1) lookups.
 *
 * What the file name defines:
 * "Blocklist" identifies the data being managed (forbidden domains).
 * "Manager" means it controls the fetching, caching, and querying of this data.
 */
public class BlocklistManager {

    private static final String TAG = "BlocklistManager";
    private static final String PREFS_NAME = "blocklist_prefs";
    private static final String KEY_LAST_UPDATE = "last_update_epoch";
    private static final long UPDATE_INTERVAL_MS = 24L * 60 * 60 * 1000; // 24 hours
    private static final String CACHE_FILE_NAME = "blocklist_cache.txt";

    private static final List<String> REMOTE_BLOCKLIST_URLS = Arrays.asList(
            "https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/porn/hosts"
            // "https://raw.githubusercontent.com/hagezi/dns-blocklists/main/hosts/porn.txt"
    );

    // Using ConcurrentHashMap.newKeySet() instead of a standard HashSet ensures
    // thread safety when the background updater and the VPN network thread access it simultaneously.
    private final Set<String> combinedBlocklist = ConcurrentHashMap.newKeySet();
    private volatile boolean isLoaded = false;
    private final Context context;

    // Replaces Kotlin Coroutines (Dispatchers.IO)
    private final ExecutorService ioExecutor = Executors.newSingleThreadExecutor();

    public BlocklistManager(Context context) {
        this.context = context.getApplicationContext();
    }

    /**
     * Initializes the blocklist. Call this when the VPN service starts.
     * In Java, we use a callback (Runnable) since we don't have Kotlin's 'suspend' keyword.
     */
    public void initialize(Runnable onComplete) {
        if (isLoaded) {
            if (onComplete != null) onComplete.run();
            return;
        }

        ioExecutor.execute(() -> {
            // Always start with the hard-coded seed list
            combinedBlocklist.addAll(PornDomainBlocklist.getDomains());
            Log.d(TAG, "Seed list loaded: " + PornDomainBlocklist.getDomains().size() + " domains");

            // Load cached remote list from disk
            loadCachedRemoteList();

            // Fetch updated remote list if enough time has passed
            if (shouldUpdateRemoteList()) {
                fetchRemoteBlocklists();
            }

            isLoaded = true;
            Log.i(TAG, "✅ Blocklist ready: " + combinedBlocklist.size() + " total domains");

            if (onComplete != null) {
                onComplete.run();
            }
        });
    }

    /**
     * Fast check if a domain should be blocked.
     */
    public boolean isBlocked(String domain) {
        if (domain == null || domain.isEmpty()) return false;

        String lower = domain.toLowerCase();

        // 1. Check exact match O(1)
        if (combinedBlocklist.contains(lower)) return true;

        // 2. Check subdomain match O(N) (e.g., "www.pornhub.com" matches "pornhub.com")
        for (String blocked : combinedBlocklist) {
            if (lower.endsWith("." + blocked)) {
                return true;
            }
        }
        return false;
    }

    public void addCustomDomain(String domain) {
        if (domain != null) {
            combinedBlocklist.add(domain.toLowerCase().trim());
            Log.d(TAG, "Custom domain added: " + domain);
        }
    }

    public void removeCustomDomain(String domain) {
        if (domain != null) {
            combinedBlocklist.remove(domain.toLowerCase().trim());
        }
    }

    // ── Private I/O Methods ───────────────────────────────────────────────────

    private void loadCachedRemoteList() {
        File file = context.getFileStreamPath(CACHE_FILE_NAME);
        if (!file.exists()) return;

        // try-with-resources automatically closes the streams to prevent memory leaks
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file)))) {
            String line;
            int count = 0;
            while ((line = reader.readLine()) != null) {
                String domain = parseHostsLine(line);
                if (domain != null) {
                    combinedBlocklist.add(domain);
                    count++;
                }
            }
            Log.d(TAG, "Loaded " + count + " domains from cache");
        } catch (Exception e) {
            Log.w(TAG, "Failed to load cached blocklist: " + e.getMessage());
        }
    }

    private void fetchRemoteBlocklists() {
        int totalFetched = 0;

        for (String urlString : REMOTE_BLOCKLIST_URLS) {
            HttpURLConnection connection = null;
            try {
                Log.d(TAG, "Fetching blocklist from: " + urlString);
                URL url = new URL(urlString);
                connection = (HttpURLConnection) url.openConnection();
                connection.setConnectTimeout(10_000);
                connection.setReadTimeout(30_000);

                Set<String> newDomains = ConcurrentHashMap.newKeySet();

                try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        String domain = parseHostsLine(line);
                        if (domain != null) {
                            newDomains.add(domain);
                        }
                    }
                }

                combinedBlocklist.addAll(newDomains);
                totalFetched += newDomains.size();

                // Cache to disk immediately
                cacheBlocklist(newDomains);
                Log.i(TAG, "Fetched " + newDomains.size() + " domains from " + urlString);

            } catch (Exception e) {
                Log.w(TAG, "Failed to fetch from " + urlString + ": " + e.getMessage());
            } finally {
                if (connection != null) connection.disconnect();
            }
        }

        if (totalFetched > 0) {
            markUpdateTime();
            Log.i(TAG, "Remote update complete: " + totalFetched + " new domains");
        }
    }

    private void cacheBlocklist(Set<String> domains) {
        try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(
                context.openFileOutput(CACHE_FILE_NAME, Context.MODE_PRIVATE)))) {
            for (String domain : domains) {
                writer.write("0.0.0.0 " + domain + "\n");
            }
        } catch (Exception e) {
            Log.w(TAG, "Failed to cache blocklist: " + e.getMessage());
        }
    }

    private String parseHostsLine(String line) {
        String trimmed = line.trim();
        if (trimmed.isEmpty() || trimmed.startsWith("#")) return null;

        String[] parts = trimmed.split("\\s+");
        if (parts.length < 2) return null;

        String ip = parts[0];
        if (!"0.0.0.0".equals(ip) && !"127.0.0.1".equals(ip)) return null;

        String domain = parts[1].toLowerCase();

        // Skip localhost and invalid entries
        if ("localhost".equals(domain) || domain.contains("#") || domain.length() < 4) return null;

        return domain;
    }

    private boolean shouldUpdateRemoteList() {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        long lastUpdate = prefs.getLong(KEY_LAST_UPDATE, 0L);
        return System.currentTimeMillis() - lastUpdate > UPDATE_INTERVAL_MS;
    }

    private void markUpdateTime() {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                .edit()
                .putLong(KEY_LAST_UPDATE, System.currentTimeMillis())
                .apply();
    }
}

File: app/src/main/java/com/example/digitalmonk/service/vpn/blocklist/BlocklistManager.kt
package com.example.digitalmonk.service.vpn.blocklist

import android.content.Context
import android.util.Log
import com.example.digitalmonk.service.vpn.blocklist.PornDomainBlocklist
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URL

/**
 * Manages all domain blocklists.
 *
 * Sources (in priority order):
 *  1. [PornDomainBlocklist]  — hard-coded seed list, always available
 *  2. Remote blocklist files — downloaded from trusted sources (StevenBlack, etc.)
 *  3. Custom user domains    — parent-defined domains (from Room DB)
 *
 * The combined set is stored in memory for fast O(1) lookup during DNS filtering.
 *
 * ── Remote blocklist format ──────────────────────────────────────────────────
 * Standard hosts file format:
 *   0.0.0.0 pornhub.com
 *   0.0.0.0 xvideos.com
 *   # comment lines are ignored
 */
class BlocklistManager(private val context: Context) {

    private val combinedBlocklist = mutableSetOf<String>()
    private var isLoaded = false

    companion object {
        private const val TAG = "BlocklistManager"
        private const val PREFS_NAME = "blocklist_prefs"
        private const val KEY_LAST_UPDATE = "last_update_epoch"
        private const val UPDATE_INTERVAL_MS = 24L * 60 * 60 * 1000  // 24 hours

        /**
         * Trusted blocklist sources.
         * These are well-maintained, community-curated hosts files.
         *
         * StevenBlack's consolidated list:
         *   https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/porn/hosts
         *
         * This list contains 30,000+ adult domains and is updated regularly.
         */
        private val REMOTE_BLOCKLIST_URLS = listOf(
            // StevenBlack porn-only blocklist
            "https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/porn/hosts"
            // Add more sources here:
            // "https://raw.githubusercontent.com/hagezi/dns-blocklists/main/hosts/porn.txt"
        )
    }

    /**
     * Initializes the blocklist from all sources.
     * Call this when the VPN service starts.
     */
    suspend fun initialize() = withContext(Dispatchers.IO) {
        if (isLoaded) return@withContext

        // Always start with the hard-coded seed list
        combinedBlocklist.addAll(PornDomainBlocklist.domains)
        Log.d(TAG, "Seed list loaded: ${PornDomainBlocklist.domains.size} domains")

        // Load cached remote list from disk
        loadCachedRemoteList()

        // Fetch updated remote list if enough time has passed
        if (shouldUpdateRemoteList()) {
            fetchRemoteBlocklists()
        }

        isLoaded = true
        Log.i(TAG, "✅ Blocklist ready: ${combinedBlocklist.size} total domains")
    }

    /**
     * Fast O(1) check if a domain should be blocked.
     * Falls back to seed list if remote list hasn't loaded yet.
     */
    fun isBlocked(domain: String): Boolean {
        val lower = domain.lowercase()

        // Check exact match
        if (lower in combinedBlocklist) return true

        // Check subdomain match (e.g., "www.pornhub.com" → matches "pornhub.com")
        return combinedBlocklist.any { blocked ->
            lower.endsWith(".$blocked")
        }
    }

    /**
     * Adds a custom domain to the in-memory blocklist.
     * Persist this to Room DB separately for survival across restarts.
     */
    fun addCustomDomain(domain: String) {
        combinedBlocklist.add(domain.lowercase().trim())
        Log.d(TAG, "Custom domain added: $domain")
    }

    fun removeCustomDomain(domain: String) {
        combinedBlocklist.remove(domain.lowercase().trim())
    }

    // ── Private ───────────────────────────────────────────────────────────────

    private fun loadCachedRemoteList() {
        try {
            val file = context.getFileStreamPath("blocklist_cache.txt")
            if (!file.exists()) return

            var count = 0
            file.bufferedReader().forEachLine { line ->
                val domain = parseHostsLine(line)
                if (domain != null) {
                    combinedBlocklist.add(domain)
                    count++
                }
            }
            Log.d(TAG, "Loaded $count domains from cache")
        } catch (e: Exception) {
            Log.w(TAG, "Failed to load cached blocklist: ${e.message}")
        }
    }

    private suspend fun fetchRemoteBlocklists() = withContext(Dispatchers.IO) {
        var totalFetched = 0

        for (url in REMOTE_BLOCKLIST_URLS) {
            try {
                Log.d(TAG, "Fetching blocklist from: $url")
                val connection = URL(url).openConnection()
                connection.connectTimeout = 10_000
                connection.readTimeout    = 30_000

                val newDomains = mutableSetOf<String>()
                connection.getInputStream().bufferedReader().forEachLine { line ->
                    val domain = parseHostsLine(line)
                    if (domain != null) newDomains.add(domain)
                }

                combinedBlocklist.addAll(newDomains)
                totalFetched += newDomains.size

                // Cache to disk
                cacheBlocklist(newDomains)

                Log.i(TAG, "Fetched ${newDomains.size} domains from $url")

            } catch (e: Exception) {
                Log.w(TAG, "Failed to fetch from $url: ${e.message}")
                // Continue with next URL — don't crash if one fails
            }
        }

        if (totalFetched > 0) {
            markUpdateTime()
            Log.i(TAG, "Remote update complete: $totalFetched new domains")
        }
    }

    private fun cacheBlocklist(domains: Set<String>) {
        try {
            context.openFileOutput("blocklist_cache.txt", Context.MODE_PRIVATE).bufferedWriter().use { writer ->
                domains.forEach { writer.write("0.0.0.0 $it\n") }
            }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to cache blocklist: ${e.message}")
        }
    }

    /**
     * Parses a single line from a hosts file.
     * Returns the domain if it's a block entry, null otherwise.
     * Format: "0.0.0.0 domain.com" or "127.0.0.1 domain.com"
     */
    private fun parseHostsLine(line: String): String? {
        val trimmed = line.trim()
        if (trimmed.isEmpty() || trimmed.startsWith("#")) return null

        val parts = trimmed.split(Regex("\\s+"))
        if (parts.size < 2) return null

        val ip = parts[0]
        if (ip != "0.0.0.0" && ip != "127.0.0.1") return null

        val domain = parts[1].lowercase()

        // Skip localhost and invalid entries
        if (domain == "localhost" || domain.contains("#") || domain.length < 4) return null

        return domain
    }

    private fun shouldUpdateRemoteList(): Boolean {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val lastUpdate = prefs.getLong(KEY_LAST_UPDATE, 0L)
        return System.currentTimeMillis() - lastUpdate > UPDATE_INTERVAL_MS
    }

    private fun markUpdateTime() {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putLong(KEY_LAST_UPDATE, System.currentTimeMillis())
            .apply()
    }
}